### PR TITLE
Training continuation

### PR DESCRIPTION
**Pull request by gabrielBusta**
This is work around #154 

I iterated a few times so there's some fluff, and I am not sure how to deal with the vocab.

There's an unused wrapper script called `taskcluster/scripts/pipeline/spm-vocab-taskcluster.sh`

I added that script originally because I thought we only needed to replace the `train-vocab` step on the graph. But actually the vocabulary is specific to each ensemble/model. That means it needs to be specified per pre trained model. So I made the vocab url a part of the pre trained model config for the teacher ensemble. Similarly, when we add continuation to the backward model, we need to specify the vocabulary for that particular backwards model.

On this instance, that means that the `train-teacher` kind (when it is being continued) doesn't really depend on the `train-vocab` step that executed on the pipeline.

I configure the pre trained models in the action input like this:

```
datasets:
  devtest:
    - flores_dev
    - sacrebleu_wmt19
  mono-src:
    - news-crawl_news.2008
  mono-trg:
    - news-crawl_news.2007
  test:
    - flores_devtest
  train:
    - opus_ada83/v1
    - opus_ELRC-3075-wikipedia_health/v1
experiment:
  backward-model: NOT-YET-SUPPORTED
  best-model: chrf
  bicleaner:
    dataset-thresholds:
      opus_ELRC-3075-wikipedia_health/v1: 0.6
      opus_ada83/v1: 0.0
    default-threshold: 0.5
  mono-max-sentences-src: 10000
  mono-max-sentences-trg: 10000
  name: ci
  split-length: 5000
  spm-sample-size: 10000
  spm-vocab-size: 1000
  src: ru
  teacher-ensemble: 2
  trg: en
  use-opuscleaner: 'true'
  vocab: NOT-YET-SUPPORTED
  pretrained-models:
    teacher-base:
      urls:
        - "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/YQUp-mUgTNSsJS6RjVSUrQ/artifacts/public/build"
        - "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/bpUaubrpQgOPegv8ZH9xTg/artifacts/public/build"
      mode: use
      type: default
      vocab:
        url: "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/Vof1ufPxTOyG3OiSj4oHwQ/artifacts/public/build"
marian-args:
  decoding-backward:
    mini-batch-words: '2000'
  decoding-teacher:
    mini-batch-words: '1000'
    precision: float16
  training-backward:
    after: 10u
    dim-vocabs: 1000 1000
    disp-freq: '1'
    save-freq: '5'
    valid-freq: '10'
  training-student:
    after: 10u
    dim-vocabs: 1000 1000
    disp-freq: '1'
    save-freq: '5'
    valid-freq: '10'
  training-student-finetuned:
    after: 10u
    dim-vocabs: 1000 1000
    disp-freq: '1'
    save-freq: '5'
    valid-freq: '10'
  training-teacher-base:
    after: 20u
    dim-vocabs: 1000 1000
    disp-freq: '1'
    save-freq: '5'
    task: transformer-base
    valid-freq: '10'
  training-teacher-finetuned:
    after: 10u
    dim-vocabs: 1000 1000
    disp-freq: '1'
    save-freq: '5'
    task: transformer-base
    valid-freq: '10'
target-stage: train-teacher
taskcluster:
  split-chunks: 2
```

### CODE REVIEW

**Comment by gabrielBusta** [comment]
Blah maybe I do need the wrapper around vocab? I started to see this dataset mismatches https://firefox-ci-tc.services.mozilla.com/tasks/GjNToUN5TLOsiRPREl3Ing/runs/0/logs/public/logs/live.log

**Comment by gabrielBusta** [comment]
Ah it's probably because translate mono depends on train-vocab :/

**Comment by gabrielBusta** [comment]
@bhearsum there's some [additional context here](https://docs.google.com/document/d/1fG2YK6ZttBFVZMWHuZjwunc-jdwmN2rBvokoXhgs1bQ/edit#heading=h.jotjwzta1ziy)

**Comment by gabrielBusta** [comment]
Does the backward model have to use the same vocab as the teacher ensemble @eu9ene? I thought they could be different

**Comment by gabrielBusta** [comment]
I also had a "Continue Training" action (more or less a re-trigger but it mounts the artifacts from the task it continues from) but I decided to take out because it would be tricky to deal with edge cases (what happens if you "Continue Training" a task that was in "init" mode for example? The `--pretrained-model` flag would affect the continuation, when it should actually behave like "continue".) Also, not sure how I could simulate a spot outage to actually validate it. So I stashed it.

**Comment by eu9ene** [comment]
@gabrielBusta great job! We're getting close.

> Does the backward model have to use the same vocab as the teacher ensemble @eu9ene? I thought they could be different

Yes, we can use pre-trained teachers and backward models that were trained with different vocabs.

The vocab should basically become a part of the model. It should be stored in the same directory, as a model. It should be copied before training if we train it as a part of the pipeline or copied together with the other model artifacts in pre-trained mode.

For "translate" steps we should use the vocab from the train step artifacts and not the one we trained separately because it might be different for the pre-trained models. 

I like the idea of pulling the pre-trained vocab too if needed, but I would remove the vocab URLs from the config for the pre-trained models and just retrieve it together with the model artifacts. It would be less error-prone and simplify usage.

**Comment by gabrielBusta** [comment]
I see. To echo the idea (now that we discussed it more.) We can make the vocabulary a model was trained on an artifact of the training task. That way we can omit `pretrained-models.teacher-base.vocab.url` (we'd get the vocabulary as an artifact stored along-side the model.) Tasks where the vocabulary must be the one that is bundled with the model could depend on the training task and use the vocabulary artifact there. The training task will use the spm vocab from `train-vocab` if it's not a continuation. Tasks where we want to use the vocabulary trained in the pipeline can depend on `train-vocab` (for example the `alignment` tasks)

**Comment by bhearsum** [comment]
> The vocab should basically become a part of the model. It should be stored in the same directory, as a model. It should be copied before training if we train it as a part of the pipeline or copied together with the other model artifacts in pre-trained mode.
> 
> For "translate" steps we should use the vocab from the train step artifacts and not the one we trained separately because it might be different for the pre-trained models.

:grimacing: - sounds like this is another bug in the original Taskcluster pipeline (we're using using vocab from `train-vocab` everywhere as far as I know). It looks like that will get fixed here though?

> I like the idea of pulling the pre-trained vocab too if needed, but I would remove the vocab URLs from the config for the pre-trained models and just retrieve it together with the model artifacts. It would be less error-prone and simplify usage.

Just to make sure I understand: the idea is that we would simply assume it exists next to any pretrained model we point at? (Or would we point at a taskid instead, or...?)

**Comment by eu9ene** [comment]
> 😬 - sounds like this is another bug in the original Taskcluster pipeline (we're using using vocab from train-vocab everywhere as far as I know). It looks like that will get fixed here though?

It's not a bug. We use the trained vocab everywhere now. When we implement training continuation, it should be flexible enough to use a bunch of different pre-trained models with different vocabs, so it makes sense to make the vocab a part of the model artifacts to make sure we use the right one for each model.

> Just to make sure I understand: the idea is that we would simply assume it exists next to any pretrained model we point at? (Or would we point at a taskid instead, or...?)

Yes, the vocab that we use to train the model should be copied to the model output artifacts. When we use a pre-trained model we should assume it's in the folder with the other artifacts. We can still reuse the pre-trained vocab if we implement pulling it by URL and skipping the vocab training step, but it's optional.

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:55]: This transform's name (`training_continuation`) suggests that it could be used on any training step, but some of the details such as this make it clear it only works for `train-teacher` right now. It would be lovely if it could be made generic enough to work with any training step (eg: by parameterizing some things). If not, it ought to be renamed.

(It can probably be refactored in such a way that most of the code can be re-used even if we maintain separate transforms for each training step.)

File: taskcluster/translations_taskgraph/parameters.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/parameters.py:24]: I will call out that removing these means that we're diverging further from the snakemake configuration. That's probably OK and necessary for this case - and perhaps a sign that keeping up the pretense of having compatible configs is not worthwhile at this point.

File: taskcluster/translations_taskgraph/parameters.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/parameters.py:156]: This raises a question: are we going to support testing pretrained model support in the quick CI runs? Currently we don't, as these are absent from the defaults above. (If we are, that could totally be added in a follow-up, or filed as an issue for the future.)

File: taskcluster/ci/train-backwards/kind.yml
Comment (bhearsum) [inline taskcluster/ci/train-backwards/kind.yml:44]: I'm assuming this is just not quite complete yet? (Some of these arguments are being added, but the `training_continuation` transform is not being used yet.)

**Review by bhearsum — COMMENTED** [review]
This work looks very promising!

There's seems to some issues with the new vocab getting saved -- I'm guessing that there's a `cd` that happens at some point which invalidates the relative path, but I haven't verified that. You should probably ensure there's an absolute path in the `train-taskcluster.sh` wrapper regardless (unless there's a env var we can use to get the absolute path to the artifacts dir).

> [task 2023-10-19T01:24:54.154Z] trainer_interface.cc(605) LOG(INFO) Saving model: ./artifacts/vocab.spm.model
> [task 2023-10-19T01:24:54.197Z] [2023-10-19 01:24:54] Error: SentencePiece vocabulary error: Permission denied: "./artifacts/vocab.spm.model": No such file or directory Error #2
[task 2023-10-19T01:24:54.197Z] [2023-10-19 01:24:54] Error: Aborted from virtual void marian::SentencePieceVocab::create(const string&, const 

I think I could use a quick explainer on why we're using mounts to implement this as well. Is it just to let generic-worker deal with the downloading? AFAICT we don't save on any work compared to having our own script fetch. (In fact, it might even be slower if large files are involved, because the logs appear to show at least one copy being made:
> [taskcluster 2023-10-19T01:24:38.573Z] [mounts] Copying /home/ubuntu/downloads/Dkb7WdruTIyAPyekGV_dzw to /home/ubuntu/tasks/task_169767843132649/artifacts/vocab.spm

**Comment by bhearsum** [comment]
> > Just to make sure I understand: the idea is that we would simply assume it exists next to any pretrained model we point at? (Or would we point at a taskid instead, or...?)
> 
> Yes, the vocab that we use to train the model should be copied to the model output artifacts. When we use a pre-trained model we should assume it's in the folder with the other artifacts. We can still reuse the pre-trained vocab if we implement pulling it by URL and skipping the vocab training step, but it's optional.

Got it (I see there's a large list of other files in the new transform as well), and that we're pointing at a directory (not a specific file) in the new training configuration -- makes sense!

File: taskcluster/translations_taskgraph/parameters.py
Comment (gabrielBusta) [reply taskcluster/translations_taskgraph/parameters.py:24]: Yeah. I went even further and used the kind names in the configuration to simplify the code in the transform

File: taskcluster/ci/train-backwards/kind.yml
Comment (gabrielBusta) [reply taskcluster/ci/train-backwards/kind.yml:45]: Yup. It is going through the transform now.

File: taskcluster/translations_taskgraph/parameters.py
Comment (gabrielBusta) [reply taskcluster/translations_taskgraph/parameters.py:24]: I see there's a `taskcluster` key in the config for `Taskcluster-specific pipeline configuration`. Maybe I can put these configuration fields in there.

File: taskcluster/translations_taskgraph/parameters.py
Comment (gabrielBusta) [reply taskcluster/translations_taskgraph/parameters.py]: Is it possible to do that, but run the usual training if the training task was modified? I could trigger an action on the  push decision.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh]: a transform puts them in there

File: taskcluster/translations_taskgraph/parameters.py
Comment (gabrielBusta) [reply taskcluster/translations_taskgraph/parameters.py]: I like the idea...

**Comment by gabrielBusta** [comment]
> I think I could use a quick explainer on why we're using mounts to implement this as well. Is it just to let generic-worker deal with the downloading?

**Comment by gabrielBusta** [comment]
> I think I could use a quick explainer on why we're using mounts to implement this as well. Is it just to let generic-worker deal with the downloading?

We could pass down the url and do the downloading in the script. Maybe with a transform that modifies the run command to insert the url for the model we want to use? Or maybe add an environment variable that the script can read? Mounting the files was straight forward 🤷 maybe the wrapper could (should?) be a Python script.

File: taskcluster/translations_taskgraph/parameters.py
Comment (bhearsum) [reply taskcluster/translations_taskgraph/parameters.py:24]: If we're diverging in such big ways I'm not sure we need to worry much about separating the Taskcluster specific things to be honest. @eu9ene - do you have an opinion here?

File: taskcluster/translations_taskgraph/parameters.py
Comment (bhearsum) [reply taskcluster/translations_taskgraph/parameters.py]: I imagine it would be possible to dynamically choose whether or not to use a pretrained X.

The downside to doing it with an action task is that it can only be done by those without specific permissions. Not the end of the world, I guess, as anyone who could review/merge presumably has those permissions...

**Comment by bhearsum** [comment]
> > I think I could use a quick explainer on why we're using mounts to implement this as well. Is it just to let generic-worker deal with the downloading?
> 
> We could pass down the url and do the downloading in the script. Maybe with a transform that modifies the run command to insert the url for the model we want to use? Or maybe add an environment variable that the script can read? Mounting the files was straight forward 🤷 maybe the wrapper could (should?) be a Python script.

I don't really have a strong opinion here - if Taskcluster/generic-worker can do it, that's great! I'm just not sure if I'm missing something. When I see `mounts` I guess I expect some sort of disk image, but it sounds like when a `mount` is a task artifact it really just means "download this thing" (which...is not really mounting it...which is why I'm double checking).

**Comment by gabrielBusta** [comment]
@bhearsum ah OK. I think about it more as "downloads this thing." I came across them modifying [the `mounts` that download `run-task` and `fetch-content`](https://searchfox.org/mozilla-central/source/taskcluster/gecko_taskgraph/transforms/job/run_task.py#189-205) on [Bug 1760675](https://bugzilla.mozilla.org/show_bug.cgi?id=1760675).

**Comment by gabrielBusta** [comment]
tbh, I did consider refactoring into a (Python) script (and passing the config through the task environment or command argument.) It seems like it would be more flexible. But the `mounts` are at a point now where we can get value out of this patch soon. So, I was thinking of working out the kinks and landing something that can be used for the continuation of `train-teacher` and `train-backwards` (and `finetune-teacher`?)

**Comment by gabrielBusta** [comment]
I could do some refactoring now, but me and @eu9ene were chatting earlier this week about this and #161 / #219 and it seems once the continuation and opus trainer land, it will be a good time to re-group.

**Comment by eu9ene** [comment]
> I could do some refactoring now, but me and @eu9ene were chatting earlier this week about this and #161 / #219 and it seems once the continuation and opus trainer land, it will be a good time to re-group.

Yes, I'd prefer not to do a major refactoring of the training part at this point. Let's land OpusTrainer and this with the current approach and when we have everything working let's think how to reduce complexity. We'll likely need it to integrate pre-trained OPUS-MT models.

**Comment by gabrielBusta** [comment]
@eu9ene could you try this out? I've been running into issues, but I think some of it might be configuring things in ways that don't reflect actual use cases. Can you do basic continuation/use with the teacher and backward models?

**Comment by gabrielBusta** [comment]
@eu9ene do you know what this could be about?

```
[task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Requested shape shape=1000x512 size=512000 for existing parameter 'Wemb' does not match original shape shape=32000x1024 size=32768000
[task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Requested shape shape=1000x512 size=512000 for existing parameter 'Wemb' does not match original shape shape=32000x1024 size=32768000
[task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Aborted from marian::Expr marian::ExpressionGraph::param(const string&, const marian::Shape&, marian::Ptr<marian::inits::NodeInitializer>&, marian::Type, bool, bool) in /builds/worker/fetches/marian-source/src/graph/expression_graph.h:410
```

https://firefox-ci-tc.services.mozilla.com/tasks/EFSrTdfiTEe17M3qyFE3hw/runs/0/logs/public/logs/live.log#L576-578

(it works fine with the taskcluster model)

**Comment by eu9ene** [comment]
> @eu9ene do you know what this could be about?
> 
> ```
> [task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Requested shape shape=1000x512 size=512000 for existing parameter 'Wemb' does not match original shape shape=32000x1024 size=32768000
> [task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Requested shape shape=1000x512 size=512000 for existing parameter 'Wemb' does not match original shape shape=32000x1024 size=32768000
> [task 2023-11-14T00:01:33.671Z] [2023-11-14 00:01:33] Error: Aborted from marian::Expr marian::ExpressionGraph::param(const string&, const marian::Shape&, marian::Ptr<marian::inits::NodeInitializer>&, marian::Type, bool, bool) in /builds/worker/fetches/marian-source/src/graph/expression_graph.h:410
> ```
> 
> https://firefox-ci-tc.services.mozilla.com/tasks/EFSrTdfiTEe17M3qyFE3hw/runs/0/logs/public/logs/live.log#L576-578
> 
> (it works fine with the taskcluster model)

It's probably because this model was trained with different marian parameters

**Comment by eu9ene** [comment]
> How do you know what parameters it was trained on?

You can look at the train.log. Marian outputs the parameters with `[config]` prefix.

File: taskcluster/translations_taskgraph/parameters.py
Comment (eu9ene) [reply taskcluster/translations_taskgraph/parameters.py:24]: It's fine, I have separated example configs for Snakemake and TC anyway

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (eu9ene) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:8]: It's fine to have this list but there might be more metrics used, for example OPUS-MT use perplexity

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (eu9ene) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:24]: There should be final.* name. It's the file we copy in the end which might not be the case for older models. Also only one model file is required for initialization.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (eu9ene) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:37]: again it's fine to use with the current pipeline, but  theoretically `best_model_metric` setting can be different in those runs, so we can just copy final.model.npz.best-*.npz

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (eu9ene) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:43]: why don't we init from the same model we copy to use? final.*

File: taskcluster/ci/finetune-teacher/kind.yml
Comment (eu9ene) [inline taskcluster/ci/finetune-teacher/kind.yml:1]: this step will go away after OpusTrainer is landed

**Comment by eu9ene** [comment]
It would be great to add docs on how to use it, structure of the configs and behavior of each mode.

File: taskcluster/ci/evaluate-teacher-ensemble/kind.yml
Comment (bhearsum) [inline taskcluster/ci/evaluate-teacher-ensemble/kind.yml:119]: praise: Thank you for cleaning this up here, and elsewhere. It's nice to avoid the unnecessary direct dependency on `train-vocab`.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:26]: nit: It's odd to be comparing against `None` here. Usually in bash you expect an empty or undefined string as a null value.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:29]: question: I want to make sure I'm understanding this correctly. If we're using a pretrained model, we pull the vocab from a mount? And if not, we pull it from a fetch?

That all seems fine, but if this is true, I would prefer to see the fetch and dependency on `train-vocab` not happening at all if we're using a pretrained model. It's not a large file, but it makes it much less clear where the vocab comes from if we still maintain this dependency while not using anything from it. Please make sure there's a sanity check that ensure that either a pretrained model being used OR a dependency on `train-vocab` in this case.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:7]: todo: Replace this with https://github.com/mozilla/firefox-translations-training/blob/0e757b0070a98599d4ade0df01eb85e131718d43/taskcluster/scripts/pipeline/translate-taskcluster.sh#L11 (there's no need to require callers to pass this in - you can figure it out more reliably in the script).

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:1]: question: Should we be supporting pretrained models for student training as well? (This is probably a question for @eu9ene).

(If we should, I suggest this is a future enhancement, and not something for this PR.)

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:35]: question: what is the use case for the `use` mode? If we already have models that we want to use, can we not just start training at a later phase, and pick them up as a cached task?

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:41]: question: is there an explainer for why we need both of these modes? I see they mount different artifacts, but it's not clear to me what circumstances we would use them both in.

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train-taskcluster.sh:1]: praise: I think it was a great idea to handle all of the training continuation bits in this script instead of trying to have `train.sh` support it in addition to its core job.

idea: Is there a use case for supporting training continuation outside of Taskcluster? If so, we may want to get an issue on file for making this script Taskcluster agnostic. (I explicitly don't want to add that work to this PR even if it may be useful.)

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:31]: nitpick: It may be better to run this through `urlparse` rather than try to cleanup the url yourself.

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:53]: suggestion (non-blocking): This looks like it's purely validating parameters. If so, this is probably better to do when we first see them in `train.py`.

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:75]: nitpick: `add_pretrained_model_mounts` would be a better name. The current name suggests that this function is actually mounting something, which is obviously not the case.

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/training_continuation.py:88]: suggestion (blocking): Use `config.kind` here. (The metadata is not guaranteed to be present.)

**Review by bhearsum — CHANGES_REQUESTED** [review]
I think this looks pretty reasonable overall. I agree with @eu9ene's sentiment that some docs and further explanation would be useful. I have a few specific questions in this regard further down.

I know this needs rebasing, but if I understand correctly the main change will be that all of the `finetune-teacher` references will simply use `train-teacher` instead - which I've assumed while doing this review.

Also, I'm trying to use https://conventionalcomments.org/#labels here, hopefully it helps!

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (eu9ene) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:1]: Snakemake already supports this because it works with files. This work is Taskcluster specific, so this looks like a good place to handle it.

**Review by eu9ene — APPROVED** [review]
I'll be unavailable so approving now. It looks fine overall. I'm ok with merging after addressing the issues with the "init" mode and "final*" models. + @bhearsum's comments.

**Comment by gabrielBusta** [comment]
> I think this looks pretty reasonable overall. I agree with @eu9ene's sentiment that some docs and further explanation would be useful. I have a few specific questions in this regard further down.
> 
> I know this needs rebasing, but if I understand correctly the main change will be that all of the `finetune-teacher` references will simply use `train-teacher` instead - which I've assumed while doing this review.
> 
> Also, I'm trying to use https://conventionalcomments.org/#labels here, hopefully it helps!

@bhearsum there was a little more to it because of the alignments and evaluate stages. CI is back to healthy now

**Comment by bhearsum** [comment]
> > I think this looks pretty reasonable overall. I agree with @eu9ene's sentiment that some docs and further explanation would be useful. I have a few specific questions in this regard further down.
> > I know this needs rebasing, but if I understand correctly the main change will be that all of the `finetune-teacher` references will simply use `train-teacher` instead - which I've assumed while doing this review.
> > Also, I'm trying to use https://conventionalcomments.org/#labels here, hopefully it helps!
> 
> @bhearsum there was a little more to it because of the alignments and evaluate stages. CI is back to healthy now

It looks like you may have missed many of my earlier review comments - can you take a look at those before I review this again, please?

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:30]: Mm I can probably do something regarding that in training_continuation.py

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:1]: Are we handling the spot terminations gracefully in Taskcluster now? Last time I tried to continue from a preempted VM I couldn't because the worker didn't fail gracefully and lost the checkpoint artifacts.

File: taskcluster/translations_taskgraph/transforms/training_continuation.py
Comment (gabrielBusta) [reply taskcluster/translations_taskgraph/transforms/training_continuation.py:34]: Even when I run it through `urlparse` if the input is missing that `/` there are issues using `urljoin` :/

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:39]: We want to be able to load and use models from external archives

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:43]: `continue` is "pick up where we left off using the checkpoint" and `init` means start from the beginning but use the model provided as the base model (ie pass in the `--pretrained-model` flag to `marian`)

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:33]: I refactored the code so we don't have the fetch and dependency on `train-vocab` if the `vocab.spm` file is mounted.

**Comment by gabrielBusta** [comment]
I think I will do the doc / explainer on a separate [skip CI] PR

**Review by bhearsum — APPROVED** [review]
This looks good to me - nice work!!

File: taskcluster/scripts/pipeline/train-taskcluster.sh
Comment (gabrielBusta) [reply taskcluster/scripts/pipeline/train-taskcluster.sh:43]: Yeah, it's for fine-tuning/refining a pre-existing model

**Pull request merged**
